replit_river/server_transport.py

```diff
@@ -148,30 +146,74 @@ async def _establish_handshake(
                 websocket,
             )
             raise InvalidMessageException("handshake request to wrong server")
-        if handshake_request.expectedSessionState is None:
-            # TODO: remove once we have upgraded all clients
-            my_session_id = await self._get_or_create_session_id(
-                request_message.from_, handshake_request.sessionId
+
+        old_session = self._sessions.get(request_message.from_, None)
+        if old_session and old_session.session_id == handshake_request.sessionId:
+            # check invariants
+
+            # ordering must be correct
+            client_next_expected_seq = (
+                handshake_request.expectedSessionState.nextExpectedSeq
             )
-        elif handshake_request.expectedSessionState.reconnect:
-            maybe_my_session_id = await self._get_existing_session_id(
-                request_message.from_,
-                handshake_request.sessionId,
-                handshake_request.expectedSessionState.nextExpectedSeq,
+            client_next_sent_seq = (
+                handshake_request.expectedSessionState.nextSentSeq or 0
             )
-            if maybe_my_session_id is None:
-                handshake_response = await self._send_handshake_response(
+            our_next_seq = await old_session.get_next_sent_seq()
+            our_ack = await old_session.get_next_expected_seq()
+
+            if client_next_sent_seq > our_ack:
+                message = (
```

Comment:
Should we close & delete the old session here?

Reply:
no, we tell the client to go away in this case to start a new session

Reply:
we dont want to delete sessions if we dont need to and would rather wait for the session to gracefully die